framework/test/gtest/ServiceFactoryTest.cpp

```diff
@@ -201,3 +232,389 @@ TEST_F(ServiceFactoryTest, TestGetServiceReturnsNull)
+  auto svcGetServiceThrowsRefs(context.GetServiceReferences(
+    "cppmicroservices::TestBundleH", getServiceThrowsFilter));
+  //Test that Number of service references returned is 1.
+  ASSERT_EQ(svcGetServiceThrowsRefs.size(), 1);
+  //Test that 'getservice_exception' service property is 'true'
+  ASSERT_EQ("1",
+            svcGetServiceThrowsRefs[0]
+              .GetProperty(std::string("getservice_exception"))
+              .ToString());
+  //Test that the service object returned is a nullptr
+  ASSERT_EQ(nullptr, context.GetService(svcGetServiceThrowsRefs[0]));
+  //Test that one FrameworkEvent was sent
+  ASSERT_EQ(1, fwEvents.size());
+
+  bundleH.GetBundleContext().RemoveListener(std::move(listenerToken));
+
+  fwEvents.clear();
+  listenerToken =
+    bundleH.GetBundleContext().AddFrameworkListener(eventCountListener);
+
+  std::string unGetServiceThrowsFilter(LDAPProp("ungetservice_exception") ==
+                                       true);
+  auto svcUngetServiceThrowsRefs(context.GetServiceReferences(
+    "cppmicroservices::TestBundleH", unGetServiceThrowsFilter));
+  //test that Number of service references returned is 1.
+  ASSERT_EQ(svcUngetServiceThrowsRefs.size(), 1);
+  //Test that 'ungetservice_exception' service property is 'true'
+  ASSERT_EQ("1",
+            svcUngetServiceThrowsRefs[0]
+              .GetProperty(std::string("ungetservice_exception"))
+              .ToString());
+  {
+    auto sfUngetServiceThrowSvc =
+      context.GetService(svcUngetServiceThrowsRefs[0]);
+  } // When sfUngetServiceThrowSvc goes out of scope, ServiceFactory::UngetService should be called
+  //Test that one FrameworkEvent was sent
+  ASSERT_EQ(1, fwEvents.size());
+
+  fwEvents.clear();
+  std::string recursiveGetServiceFilter(LDAPProp("getservice_recursion") ==
+                                        true);
+  auto svcRecursiveGetServiceRefs(context.GetServiceReferences(
+    "cppmicroservices::TestBundleH", recursiveGetServiceFilter));
+  //Test that Number of service references returned is 1.
+  ASSERT_EQ(svcRecursiveGetServiceRefs.size(), 1);
+  //Test that 'getservice_recursion' service property is 'true'
+  ASSERT_EQ("1",
+            svcRecursiveGetServiceRefs[0]
+              .GetProperty(std::string("getservice_recursion"))
+              .ToString());
+#if !defined(US_ENABLE_THREADING_SUPPORT) || defined(US_HAVE_THREAD_LOCAL)
+  //Test that the service object returned is a nullptr
+  ASSERT_EQ(nullptr, context.GetService(svcRecursiveGetServiceRefs[0]));
+  //Test that one FrameworkEvent was sent
+  ASSERT_EQ(2, fwEvents.size());
+  //Test for correct framwork event type
+  ASSERT_EQ(fwEvents[0].GetType(), FrameworkEvent::FRAMEWORK_ERROR);
+  //Test for correct framwork event bundle
+  ASSERT_EQ(fwEvents[0].GetBundle(), context.GetBundle());
+
+  //Test for correct service exception type (recursion)
+  EXPECT_THROW(std::rethrow_exception(fwEvents[0].GetThrowable()),
+               cppmicroservices::ServiceException);
```

Comment:
there is a missing test point for :`

`US_TEST_CONDITION_REQUIRED(
        exc.GetType() == ServiceException::FACTORY_RECURSION,`